File: 0124 Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// Don't need to check root at the very beginning since it's guaranteed non-empty.
class Solution {
    int res;
    public int maxPathSum(TreeNode root) {
        res = root.val;
        dfs(root);
        return res;
    }
    public int dfs(TreeNode root) {
        int m = root.val;
        int l = (root.left != null) ? dfs(root.left) : 0;
        int r = (root.right != null) ? dfs(root.right) : 0;
        m = Math.max(m, l + root.val);
        m = Math.max(m, r + root.val);
        res = Math.max(res, m);
        res = Math.max(res, l + r + root.val);
        return m;
    }
}

File: 0124 Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.res = 0
    
    def maxPathSum(self, root: TreeNode) -> int:        
        self.res = root.val
        self.dfs(root)
        return self.res
    
    def dfs(self, n: TreeNode) -> int:
        l = self.dfs(n.left) if n.left else 0
        r = self.dfs(n.right) if n.right else 0
        m = max(n.val, l + n.val, r + n.val)
        self.res = max(self.res, m, l + r + n.val)
        return m